crash after scene about capturing northeast sam site

The game crashed after showing the animation about northeast SAM site being captured.


```
...
2018/06/28 22:02:48 WARNING [Animations      ] Animation Surface for Body 1, animation STANDING, surface 152 not loaded.
terminate called after throwing an instance of 'std::runtime_error'
  what():  Palette creation failed, soldier has invalid animation
Creating an emergency savegame succeeded (error.sav).
Please report this error with a description of the circumstances. Do not forget to attach the savegame.
Neúspěšně ukončen (SIGABRT)
```


| JA2S version | Game version        | Vanilla bug? | Used to work? |
| ------------ | ------------------- | ------------ | ------------- |
|  git | JA2 Gold v1.10  |    (don't know)    |    yes     |



[crashaftersamscene.zip](https://github.com/ja2-stracciatella/ja2-stracciatella/files/2238268/crashaftersamscene.zip)

I also encountered a very similar error myself. So I took the save in this issue and investigated. I am able to reproduce with cheats.

This error is about `Body 1` `BGMTHREATENSTAND`(152), which is the Big Male "Threaten-Stance". Only Magic is Big Male in this save.

## Summary

There is an "actual" anim defined by `pSolder->usAnimSurface`. But the AnimCache control uses an "expected" anim from `DetermineSoldierAnimationSurface()`. They can be different when player does something in map view when outside of merc's sector.

Then the game, in certain circumstances, will unload an actually in-use surface.


## Exact steps to reproduce

1. Enable GABBI mode first
2. Load the error save. Current sector is D15.
3. Go down tactical view in D15.
3.1. Assign Magic to On Duty
3.2. Wield the Mini-14, so now Magic has the actual anim of `BGMTHREATENSTAND`
4. Go back to Map view
4.1. Select H3 where Ira is
4.2. Now assign Magic to Repair. The expected anim is no longer `BGMTHREATENSTAND`
4.3. Teleport Ira to another SAM site
4.4. Alt+"Go To Sector", taking over the SAM site
5. Fast forward at 5 min
6. Another meanwhile scene comes up
6.1. "OK" into the meanwhile cutscene. You can cancel the Esc to cancel once the scene has started.
6.2. When scene ends, load screen for D15 appears
6.3. Runtime Error with message similar to "Could not load map file 'D15.dat': Palette creation failed, soldier has invalid animation"

## Analysis

There is an AnimCache and `usAnimState` in the merc `SOLDIERSTRUCT`. Changing to Repair outside of the merc's sector seems not to update `usAnimState`, so the Merc still have the actual `BGMTHREATENSTAND`(152) stance.

Then in Meanwhile scene, there is a call to `RebuildAllSoldierShadeTables`, which attempts to load the expected anim `BGMNOTHING_STD`(110). In the process of loading, it evicts the AnimCache, removing `BGMTHREATENSTAND`(152).

After the meanwhile, the game goes back to the last loaded sector, which is D15. It load palettes using the actual `usAnimSurface` on the soldier, which is `BGMTHREATENSTAND`(152), but the surface had been unloaded already.

## Log excerpt

_Only contains relevant lines about Animation Cache (Solder 3 is Magic):_
```
2020-07-06T12:20:38 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Hit 1 ( Soldier 0 )
2020-07-06T12:20:38 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Hit 1 ( Soldier 1 )
2020-07-06T12:20:38 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Hit 183 ( Soldier 2 )
2020-07-06T12:20:38 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Determining Bump Candidate ( Soldier 3 )
2020-07-06T12:20:38 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Slot 0 Hits 0 ( Soldier 3 )
2020-07-06T12:20:38 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Slot 1 Hits 0 ( Soldier 3 )
2020-07-06T12:20:38 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Bumping 0 ( Soldier 3 )
2020-07-06T12:21:27 [DEBUG] game\Tactical\Animation_Data.cc: Surface Database: Decrementing Usage 152 ( Soldier 3 )
2020-07-06T12:21:27 [DEBUG] game\Tactical\Animation_Data.cc: Surface Database: MercUsage: 0, Global Uasage: 0
2020-07-06T12:21:27 [DEBUG] game\Tactical\Animation_Data.cc: Surface Database: Unloading Surface: 152
2020-07-06T12:21:27 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Loading Surface 110 ( Soldier 3 )
2020-07-06T12:21:27 [DEBUG] game\Tactical\Animation_Data.cc: Surface Database: Loading 110
2020-07-06T12:21:27 [DEBUG] game\Tactical\Animation_Data.cc: Surface Database: Incrementing Usage 110 ( Soldier 3 )
...
*** MEANWHILE ***
...
2020-07-06T12:21:36 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Hit 1 ( Soldier 0 )
2020-07-06T12:21:36 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Hit 1 ( Soldier 1 )
2020-07-06T12:21:36 [DEBUG] game\Tactical\Animation_Cache.cc: Anim Cache: Hit 183 ( Soldier 2 )
2020-07-06T12:21:36 [ WARN] game\Tactical\Animation_Control.cc: Animation Surface for Body 1, animation STANDING, surface 152 not loaded.
2020-07-06T12:21:49 [ERROR] game\JAScreens.cc: Runtime Error: Could not load map file 'D15.dat': Palette creation failed, soldier has invalid animation
```

--- a/src/game/Tactical/Animation_Cache.cc
+++ b/src/game/Tactical/Animation_Cache.cc
@@ -1,14 +1,16 @@
+#include "Animation_Cache.h"
+
+#include "Animation_Control.h"
+#include "Animation_Data.h"
+#include "Logger.h"
+#include "MemMan.h"
 #include "Overhead.h"
 #include "Types.h"
-#include "Animation_Cache.h"
-#include "Animation_Data.h"
-#include "Animation_Control.h"
-#include "MemMan.h"
-#include "Logger.h"
+#include <climits>
 
 #define EMPTY_CACHE_ENTRY 65000
 
-static const UINT32 guiCacheSize = MIN_CACHE_SIZE;
+static const UINT32 guiCacheSize = DEFAULT_ANIM_CACHE_SIZE;
 
 
 void InitAnimationCache(UINT16 const usSoldierID, AnimationSurfaceCacheType* const pAnimCache)
@@ -55,8 +57,8 @@ void DeleteAnimationCache( UINT16 usSoldierID, AnimationSurfaceCacheType *pAnimC
 void GetCachedAnimationSurface(UINT16 const usSoldierID, AnimationSurfaceCacheType* const pAnimCache, UINT16 const usSurfaceIndex, UINT16 const usCurrentAnimation)
 {
 	UINT8  cnt;
-	UINT8  ubLowestIndex = 0;
-	INT16  sMostHits = (INT16)32000;
+	UINT8  ubLowestIndex = -1;
+	INT16  sMostHits = INT16_MAX;
 	UINT16 usCurrentAnimSurface;
 
 	// Check to see if surface exists already
@@ -76,8 +78,9 @@ void GetCachedAnimationSurface(UINT16 const usSoldierID, AnimationSurfaceCacheTy
 	{
 		SLOGD("Anim Cache: Determining Bump Candidate ( Soldier %d )", usSoldierID);
 
-		// Determine exisiting surface used by merc
-		usCurrentAnimSurface = DetermineSoldierAnimationSurface(&GetMan(usSoldierID), usCurrentAnimation);
+		// Determine existing surface used by merc
+		SOLDIERTYPE* pSoldier = &GetMan(usSoldierID);
+		usCurrentAnimSurface = DetermineSoldierAnimationSurface(pSoldier, usCurrentAnimation);
 		// If the surface we are going to bump is our existing animation, reject it as a candidate
 
 		// If we get here, we need to remove an animation, pick the best one
@@ -88,7 +91,14 @@ void GetCachedAnimationSurface(UINT16 const usSoldierID, AnimationSurfaceCacheTy
 
 			if ( pAnimCache->usCachedSurfaces[ cnt ] == usCurrentAnimSurface )
 			{
-				SLOGD("Anim Cache: REJECTING Slot %d EXISTING ANIM SURFACE ( Soldier %d )", cnt, usSoldierID);
+				SLOGD(ST::format("Anim Cache: REJECTING Slot {} EXISTING ANIM SURFACE ( Soldier {}, Surface {} )", cnt, usSoldierID, usCurrentAnimSurface));
+			}
+			else if (pAnimCache->usCachedSurfaces[cnt] == pSoldier->usAnimSurface)
+			{
+				// The result of DetermineSoldierAnimationSurface may be inconsistent
+				// with the actual usAnimSurface, for example when player is re-assigning
+				// mercs in strategic view.
+				SLOGD(ST::format("Anim Cache: REJECTING Slot {} IN-USE ANIM SURFACE ( Soldier {}, Surface {} )", cnt, usSoldierID, pSoldier->usAnimSurface));
 			}
 			else
 			{
@@ -100,15 +110,21 @@ void GetCachedAnimationSurface(UINT16 const usSoldierID, AnimationSurfaceCacheTy
 			}
 		}
 
-		// Bump off lowest index
-		SLOGD("Anim Cache: Bumping %d ( Soldier %d )", ubLowestIndex, usSoldierID);
-		UnLoadAnimationSurface( usSoldierID, pAnimCache->usCachedSurfaces[ ubLowestIndex ] );
+		if (ubLowestIndex != -1)
+		{
+			// Bump off lowest index
+			SLOGD("Anim Cache: Bumping %d ( Soldier %d )", ubLowestIndex, usSoldierID);
+			UnLoadAnimationSurface( usSoldierID, pAnimCache->usCachedSurfaces[ ubLowestIndex ] );
 
-		// Decrement
-		pAnimCache->sCacheHits[ ubLowestIndex ] = 0;
-		pAnimCache->usCachedSurfaces[ ubLowestIndex ] = EMPTY_CACHE_ENTRY;
-		pAnimCache->ubCacheSize--;
-
+			// Decrement
+			pAnimCache->sCacheHits[ ubLowestIndex ] = 0;
+			pAnimCache->usCachedSurfaces[ ubLowestIndex ] = EMPTY_CACHE_ENTRY;
+			pAnimCache->ubCacheSize--;
+		}
+		else
+		{
+			SLOGW("Anim Cache: No slots can be evicted");
+		}
 	}
 
 	// If here, Insert at an empty slot
@@ -125,9 +141,11 @@ void GetCachedAnimationSurface(UINT16 const usSoldierID, AnimationSurfaceCacheTy
 			pAnimCache->usCachedSurfaces[ cnt ] = usSurfaceIndex;
 			pAnimCache->ubCacheSize++;
 
-			break;
+			return;
 		}
 	}
+
+	SLOGW(ST::format("Anim Cache: Failed to find slot to load surface ( Soldier {} )", usSoldierID));
 }
 
 

--- a/src/game/Tactical/Animation_Cache.cc
+++ b/src/game/Tactical/Animation_Cache.cc
@@ -112,7 +112,7 @@ void GetCachedAnimationSurface(UINT16 const usSoldierID, AnimationSurfaceCacheTy
 
 		if (ubLowestIndex == -1)
 		{
-			SLOGW(ST::format("Anim Cache: No perferred cache slot for eviction ( Soldier {} )", usSoldierID));
+			SLOGW(ST::format("Anim Cache: No preferred cache slot for eviction ( Soldier {} )", usSoldierID));
 			ubLowestIndex = 0;
 		}
 		

--- a/src/game/Tactical/Animation_Cache.cc
+++ b/src/game/Tactical/Animation_Cache.cc
@@ -110,21 +110,21 @@ void GetCachedAnimationSurface(UINT16 const usSoldierID, AnimationSurfaceCacheTy
 			}
 		}
 
-		if (ubLowestIndex != -1)
+		if (ubLowestIndex == -1)
 		{
-			// Bump off lowest index
-			SLOGD("Anim Cache: Bumping %d ( Soldier %d )", ubLowestIndex, usSoldierID);
-			UnLoadAnimationSurface( usSoldierID, pAnimCache->usCachedSurfaces[ ubLowestIndex ] );
+			SLOGW(ST::format("Anim Cache: No perferred cache slot for eviction ( Soldier {} )", usSoldierID));
+			ubLowestIndex = 0;
+		}
+		
+		// Bump off lowest index
+		SLOGD("Anim Cache: Bumping %d ( Soldier %d )", ubLowestIndex, usSoldierID);
+		UnLoadAnimationSurface( usSoldierID, pAnimCache->usCachedSurfaces[ ubLowestIndex ] );
 
-			// Decrement
-			pAnimCache->sCacheHits[ ubLowestIndex ] = 0;
-			pAnimCache->usCachedSurfaces[ ubLowestIndex ] = EMPTY_CACHE_ENTRY;
-			pAnimCache->ubCacheSize--;
-		}
-		else
-		{
-			SLOGW("Anim Cache: No slots can be evicted");
-		}
+		// Decrement
+		pAnimCache->sCacheHits[ ubLowestIndex ] = 0;
+		pAnimCache->usCachedSurfaces[ ubLowestIndex ] = EMPTY_CACHE_ENTRY;
+		pAnimCache->ubCacheSize--;
+
 	}
 
 	// If here, Insert at an empty slot

--- a/src/game/Tactical/Animation_Cache.h
+++ b/src/game/Tactical/Animation_Cache.h
@@ -3,8 +3,7 @@
 
 #include "Types.h"
 
-#define MAX_CACHE_SIZE	20
-#define MIN_CACHE_SIZE	2
+#define DEFAULT_ANIM_CACHE_SIZE	3
 
 
 struct AnimationSurfaceCacheType

